scheds/rust/scx_mitosis/src/bpf/mitosis.bpf.c

```diff
@@ -0,0 +1,932 @@
+struct bpf_iter_css;
+struct cgroup_subsys_state;
+extern int bpf_iter_css_new(struct bpf_iter_css *it,
+			    struct cgroup_subsys_state *start,
+			    unsigned int flags) __weak __ksym;
+extern struct cgroup_subsys_state *
+bpf_iter_css_next(struct bpf_iter_css *it) __weak __ksym;
+extern void bpf_iter_css_destroy(struct bpf_iter_css *it) __weak __ksym;
+
+/*
+* Along with a user_global_seq bump, indicates that cgroup->cell assignment
+* changed
+*/
+volatile bool update_cell_assignment;
+bool draining;
+
+/*
+ * This is the main driver for reconfiguration. It only runs on CPU 0
+ * To avoid
+ */
+SEC("fentry/scheduler_tick")
+int scheduler_tick_fentry(const void *ctx)
+{
+	int cell_idx, cpu_idx;
+	struct cpu_ctx *cpu_ctx;
+	struct bpf_cpumask *cpumask;
+	struct cell_cpumask_wrapper *cell_cpumaskw;
+	struct cgroup_subsys_state *root_css, *pos;
+	struct cgroup *root_cgrp;
+
+	if (bpf_get_smp_processor_id() != 0)
+		return 0;
+
+	/*
+	 * To handle races where tasks are assigned to cells that are getting
+	 * removed, we ensure cpus dispatch from their previous cell for an entire
+	 * scheduler tick. This is a crude way of mimicing RCU synchronization.
+	 */
+	if (draining) {
+		bpf_for(cpu_idx, 0, nr_possible_cpus)
+		{
+			if (!(cpu_ctx = lookup_cpu_ctx(cpu_idx)))
+				return 0;
+
+			cpu_ctx->prev_cell = cpu_ctx->cell;
+		}
+		barrier();
+		draining = false;
+	}
+
+	if (global_seq == user_global_seq)
+		return 0;
+
+	draining = true;
+	barrier();
+	/* Iterate through each cell and create its cpumask according to what
+	   userspace says */
+	bpf_for(cell_idx, 0, MAX_CELLS)
+	{
+		cpumask = bpf_cpumask_create();
+		if (!cpumask) {
+			scx_bpf_error("Failed to create cpumask");
+			return 0;
+		}
+
+		if (!(cell_cpumaskw =
+			      bpf_map_lookup_elem(&cell_cpumasks, &cell_idx))) {
+			scx_bpf_error("Failed to find cell cpumask");
+			bpf_cpumask_release(cpumask);
+			return 0;
+		}
+
+		bpf_for(cpu_idx, 0, nr_possible_cpus)
+		{
+			u8 *u8_ptr;
+
+			if ((u8_ptr = MEMBER_VPTR(
+				     cells, [cell_idx].cpus[cpu_idx / 8]))) {
+				if (*u8_ptr & (1 << (cpu_idx % 8))) {
+					bpf_cpumask_set_cpu(cpu_idx, cpumask);
+					if (!(cpu_ctx = lookup_cpu_ctx(
+						      cpu_idx))) {
+						bpf_cpumask_release(cpumask);
+						return 0;
+					}
+
+					cpu_ctx->cell = cell_idx;
+				}
+			}
+		}
+		cpumask = bpf_kptr_xchg(&cell_cpumaskw->cpumask, cpumask);
+		if (cpumask)
+			bpf_cpumask_release(cpumask);
+	}
+
+	if (update_cell_assignment) {
+		/* Is rootcg always guaranteed to have id = 1? */
+		if (!(root_cgrp = bpf_cgroup_from_id(1))) {
+			scx_bpf_error("Could not get rootcg");
+			return 0;
+		}
+
+		root_css = &root_cgrp->self;
+
+		bpf_rcu_read_lock();
+		/* Userspace gives us the full assignment of cgroup -> cell (for cgroups
+		   that its aware of). We update the cgroup_ctx with that
+		   information. */
+		bpf_for_each(css, pos, root_css,
+			     BPF_CGROUP_ITER_DESCENDANTS_PRE)
```

Comment:
Ok, I reviewed my notes. I originally did try to build all of this to update on-demand and ran into a synchronization issue - multiple tasks (from the same cg and therefore cell) can concurrently notice their gen number is behind and updating everything could happen concurrently. I couldn't protect it all via a spinlock since I call a bunch of kfuncs.

Now I think I was also making cell updates part of all that which is where all the kfuncs happen. Looking through the code, I should be able to still update the cell via tick() but move the cg update to happen on-demand. I'll try that out